### PR TITLE
crimson: use explicit, abstract interfaces for backfill's facades.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/backfill_state.h [hunk 1]
```diff
@@ -308,6 +308,37 @@ struct BackfillState::BackfillListener {
...
+struct BackfillState::PeeringFacade {
+  virtual hobject_t earliest_backfill() const = 0;
+  virtual std::set<pg_shard_t> get_backfill_targets() const = 0;
+  virtual const hobject_t& get_peer_last_backfill(pg_shard_t peer) const = 0;
+  virtual const eversion_t& get_last_update() const = 0;
+  virtual const eversion_t& get_log_tail() const = 0;
+
+  // the performance impact of `std::function` has not been considered yet.
+  // If there is any proof (from e.g. profiling) about its significance, we
+  // can switch back to the template variant.
```
Comment (rzarzynski) [inline src/crimson/osd/backfill_state.h:323]: NOTE.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

File: src/crimson/osd/backfill_facades.h [hunk 2]
```diff
@@ -12,44 +12,42 @@ namespace crimson::osd {
...
   }
 
-  bool is_backfill_target(pg_shard_t peer) const {
+  bool is_backfill_target(pg_shard_t peer) const override {
     return peering_state.is_backfill_target(peer);
   }
   void update_complete_backfill_object_stats(const hobject_t &hoid,
-                                             const pg_stat_t &stats) {
+                                             const pg_stat_t &stats) override {
     return peering_state.update_complete_backfill_object_stats(hoid, stats);
```
Comment (ronen-fr) [inline src/crimson/osd/backfill_facades.h:47]: no need for the return

File: src/crimson/osd/backfill_state.h [hunk 3]
```diff
@@ -308,6 +308,37 @@ struct BackfillState::BackfillListener {
...
+  virtual void scan_log_after(eversion_t, scan_log_func_t) const = 0;
+
+  virtual bool is_backfill_target(pg_shard_t peer) const = 0;
+  virtual void update_complete_backfill_object_stats(const hobject_t &hoid,
+                                             const pg_stat_t &stats) = 0;
+  virtual bool is_backfilling() const = 0;
+  virtual ~PeeringFacade() {}
+};
+
+// PGFacade -- a facade (in the GoF-defined meaning) simplifying the huge
```
Comment (ronen-fr) [inline src/crimson/osd/backfill_state.h:334]: I would change the comment here to describe the specific facade

File: src/crimson/osd/pg_recovery.cc [hunk 4]
```diff
@@ -523,8 +523,8 @@ void PGRecovery::on_backfill_reserved()
   using BackfillState = crimson::osd::BackfillState;
   backfill_state = std::make_unique<BackfillState>(
     *this,
-    std::make_unique<BackfillState::PeeringFacade>(pg->get_peering_state()),
-    std::make_unique<BackfillState::PGFacade>(
+    std::make_unique<crimson::osd::PeeringFacade>(pg->get_peering_state()),
+    std::make_unique<crimson::osd::PGFacade>(
```
Comment (ronen-fr) [inline src/crimson/osd/pg_recovery.cc:527]: why do we need the cast?

Comment (rzarzynski) [reply src/crimson/osd/pg_recovery.cc:527]: `crimson::osd::BackfillState::PGFacade` is the interface, `crimson::osd::PGFacade` is the crimson's implementation of it.

File: src/crimson/osd/backfill_facades.h [hunk 5]
```diff
@@ -12,44 +12,42 @@ namespace crimson::osd {
...
+struct PeeringFacade final : BackfillState::PeeringFacade {
   PeeringState& peering_state;
 
-  decltype(auto) earliest_backfill() const {
+  hobject_t earliest_backfill() const override {
     return peering_state.earliest_backfill();
   }
 
-  decltype(auto) get_backfill_targets() const {
+  std::set<pg_shard_t> get_backfill_targets() const override {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/backfill_facades.h:22]: agreed. the inner most callee returns a reference to its member variable. see `PeeringState::get_backfill_targets()`.

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/crimson/osd/backfill_facades.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/crimson/osd/backfill_facades.h]: Yup, I missed the `const &` here. Thanks for pointing this out!

File: src/crimson/osd/backfill_state.h (hunk 3 above)
Comment (rzarzynski) [reply src/crimson/osd/backfill_state.h:334]: Makes sense. Reworked the comment.

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/backfill_state.h [hunk 6]
```diff
@@ -308,6 +308,37 @@ struct BackfillState::BackfillListener {
   virtual ~BackfillListener() = default;
 };
 
+// PeeringFacade -- a facade (in the GoF-defined meaning) simplifying
+// the interface of PeeringState. The motivation is to have an inventory
+// of behaviour that must be provided by a unit test's mock.
+struct BackfillState::PeeringFacade {
+  virtual hobject_t earliest_backfill() const = 0;
+  virtual const std::set<pg_shard_t>& get_backfill_targets() const = 0;
```
Comment (tchaikov) [inline src/crimson/osd/backfill_state.h:316]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/crimson/test_backfill.cc:188:24: error: conflicting return type specified for 'virtual std::set<pg_shard_t> BackfillFixture::PeeringFacade::get_backfill_targets() const'
  188 |   std::set<pg_shard_t> get_backfill_targets() const override {
      |                        ^~~~~~~~~~~~~~~~~~~~
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/crimson/test_backfill.cc:18:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/backfill_state.h:316:39: note: overridden function is 'virtual const std::set<pg_shard_t>& crimson::osd::BackfillState::PeeringFacade::get_backfill_targets() const'
  316 |   virtual const std::set<pg_shard_t>& get_backfill_targets() const = 0;
      |                                       ^~~~~~~~~~~~~~~~~~~~
```

might need to update `src/test/crimson/test_backfill.cc` accordingly.

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/crimson/osd/backfill_state.h (hunk 6 above)
Comment (rzarzynski) [reply src/crimson/osd/backfill_state.h:316]: Oops, updated.

File: src/crimson/osd/backfill_facades.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/crimson/osd/backfill_facades.h]: The rework recalled me why there was no `const&` here. The reason was that `BackfillFixture::PeeringFacade` stores a map `pg_shard_t -> FakeReplica` which is necessary for the sake knowing e.g. `last_backfill`. The optimization imposes introduction of the duplicative `backfill_targets_as_set` to the unittest's implementation.

**Comment by ronen-fr** [comment]
LGTM

**Review by tchaikov — APPROVED** [review]

**Pull request merged**